### DISCUSSION TITLE
Serialization/Deserialization of subclasses

### DISCUSSION THREAD

**Discussion by MikesRND** [pydantic/pydantic#3293] [Question]
Hi All,

Maybe I am bending the use case of pydantic here, but I'm trying to make a round-trip to/from JSON when using subclasses.

In the following example, the restored object contains only `MainClass` objects, not `SubclassA` and `SubclassB`  objects I want.
In addition, the subclass attributes are ignored.

It seems the class name would have to be exported as part of JSON, and used when reconstructing the object, which would also require some import context.  

Am I barking up the wrong tree?  Thanks!

```python
from pydantic import BaseModel
from typing import Dict

class MainClass(BaseModel):
    x: int

class SubclassA(MainClass):
    a: int

class SubclassB(MainClass):
    b: int

class MyCatalog(BaseModel):
    stuff: Dict[str, MainClass]

if __name__=='__main__':

    cat = MyCatalog(stuff=[])
    cat.stuff['item1'] = SubclassA(x=0, a=1)
    cat.stuff['item2'] = SubclassB(x=0, b=1)

    exported = cat.json()
    imported = MyCatalog.parse_raw(exported)

    print(cat)
    print(cat.dict())
    print(cat.json())
    print(imported)
```
Results:
```
stuff={'item1': SubclassA(x=0, a=1), 'item2': SubclassB(x=0, b=1)}
{'stuff': {'item1': {'x': 0, 'a': 1}, 'item2': {'x': 0, 'b': 1}}}
{"stuff": {"item1": {"x": 0, "a": 1}, "item2": {"x": 0, "b": 1}}}
stuff={'item1': MainClass(x=0), 'item2': MainClass(x=0)}
```

**Comment by twhughes** [comment]
Try this
```python

from pydantic import BaseModel
from typing import Dict, Literal, Union

class MainClass(BaseModel):
    x: int

class SubclassA(MainClass):
    a: int
    type: Literal['A'] = 'A'

class SubclassB(MainClass):
    b: int
    type: Literal['B'] = 'B'

class MyCatalog(BaseModel):
    stuff: Dict[str, Union[SubclassA, SubclassB]]

if __name__=='__main__':

    cat = MyCatalog(stuff=[])
    cat.stuff['item1'] = SubclassA(x=0, a=1)
    cat.stuff['item2'] = SubclassB(x=0, b=1)

    exported = cat.json()
    imported = MyCatalog.parse_raw(exported)

    print(cat)
    print(cat.dict())
    print(cat.json())
    print(imported)

```

**Reply by wudstrand** [reply]
Is there a way to handle this dynamically so I can add additional subclasses during runtime and still have the serialization and deserialization work?

**Reply by tylerflex** [reply]
Yea, I think what I just commented below should work, we had the same issue in our project and use this solution.

**Reply by tylerflex** [reply]
I'm actually not sure about adding these types dynamically, but you might be able to add the type information you need within the context of the dynamic model creation explained [here](https://pydantic-docs.helpmanual.io/usage/models/#dynamic-model-creation).  For example, create a subclass 

```py
SubClass = create_model(
    'SubClass',
    type=Literal['SubClass']
    __base__=SuperClass,
)

```

**Reply by shakfu** [reply]
Here's the `pydantic2` version of @twhughes'  solution:

```python
from pydantic import BaseModel
from typing import Dict, Literal, Union

class MainClass(BaseModel):
    x: int

class SubclassA(MainClass):
    a: int
    type: Literal['A'] = 'A'

class SubclassB(MainClass):
    b: int
    type: Literal['B'] = 'B'

class MyCatalog(BaseModel):
    stuff: Dict[str, Union[SubclassA, SubclassB]]

if __name__=='__main__':

    cat = MyCatalog(stuff={})
    cat.stuff['item1'] = SubclassA(x=0, a=1)
    cat.stuff['item2'] = SubclassB(x=0, b=1)

    exported = cat.model_dump_json()
    imported = MyCatalog.model_validate_json(exported)

    print(cat)
    print(cat.model_dump())
    print(cat.model_dump_json())
    print(imported)
```

**Comment by MikesRND** [comment]
Thank you, @twhughes!  That indeed does work, but is not quite satisfying.

The key appears to be in using the type hint that explicitly calls out the subclasses, and the Parent class *in the proper order*:

``` python
stuff: Dict[str, Union[SubclassA, SubclassB, MainClass]]
```
 
combined with adding the ```.type``` property Literal to the subclasses.

It feels a little like a back door (especially the type property), and maybe not a coding pattern I could trust to production code.  Maybe that's because I don't understand what is happening behind the scenes.
It's a shame if it doesn't work out of the box, or at least with a config parameter.
Hopefully there is a "built-in" answer someone might be able to share.  Otherwise, maybe turn this into a feature request.

**Comment by tylerflex** [comment]
@clamdad 
Yea I understand it's not very satisfying, unfortunately that's just what we're working with here. You can add this kind of method to your base model to add this type field automatically.
```py
def __init_subclass__(cls):

        value = cls.__name__
        annotation = Literal[value]

        tag_field = ModelField.infer(
            name='type',
            value=value,
            annotation=annotation,
            class_validators=None,
            config=cls.__config__,
        )
        cls.__fields__['type'] = tag_field
```

Also, sometimes I found it better to use an annotated union to be explicit about using the `'type'` field to load the correct class.

```py
stuff: Dict[str, Union[SubclassA, SubclassB, MainClass]]
```

becomes

```py
from typing_extensions import Annotated
stuff: Dict[str, Annotated[Union[SubclassA, SubclassB, MainClass], pydantic.Field(discriminator='type')]]
```

Hope it helps

**Reply by shakfu** [reply]
Hi @tylerflex 

Is there a pydantic2 equivalent of:

```python
def __init_subclass__(cls):

        value = cls.__name__
        annotation = Literal[value]

        tag_field = ModelField.infer(
            name='type',
            value=value,
            annotation=annotation,
            class_validators=None,
            config=cls.__config__,
        )
        cls.__fields__['type'] = tag_field
```

**Reply by tylerflex** [reply]
Not that I could figure out, sorry. when we ported to v2, I just imported `pydantic.v1` due to complications like this. I remember trying a few options, but nothing worked as expected.  If you figure something out, let me know!

**Reply by shakfu** [reply]
Thanks for your response, yet the api completely changed..  but if I figure it out, will let you know 😄

**Reply by samuelcolvin** [reply]
There's nothing directly equivalent.

What are you trying to do?

**Reply by shakfu** [reply]
@samuelcolvin 

I'm really looking for is a way to customize the deserialization process where subclasses are involved as per my [project](https://github.com/shakfu/py2max/tree/pydantic2)

In my particular case, there are literally hundreds of object types from a c++ third-party application. These objects are serialized by the application and saved as json. 

My project allows for offline manipulation or creation of said objects by either deserializing them to python from json or creating the json from python , then deserialize them back to the json format so they can be loaded in the third party application.

While I can have one general base class to capture all of said objects types, the programming api is improved if I subclass the base class. the problem is that I am unlikely to be able to create a subclass for each object type all at once, and there needs to be a fallback to the base class if the field discriminator is somehow not linked to a subclass implementation.

What would be ideal is some kind of pydantic support for a method like the following

```python

IMPLEMENTED_SUBCLASSES = ['suba']

class Base(BaseModel):
    ...
    type: Literal['base'] = 'base'

class CategoryA(Base):
    @model_deserializer
    def deserialize(cls, data: dict) -> 'Base':
        """Customize deserialize from dict -> class"""
        if data['type'] not in IMPLEMENTED_SUBCLASSES:
            return Base(**dict)
        else:
            return cls(**dict)

class SubclassA(CategoryA):
    type: Literal['suba'] = 'suba'

```

**Comment by samuelcolvin** [comment]
See #4456 for a discussion of serialization in v2, best to comment there if you have an idea of what you want.

**Comment by metheoryt** [comment]
SerializeAsAny might be the way here https://docs.pydantic.dev/latest/concepts/serialization/#serializing-with-duck-typing

**Reply by leotac** [reply]
this does it!

**Reply by C0DK** [reply]
Doesnt help for deserialization right?

**Reply by Viicos** [reply]
@C0DK see the discussion in https://github.com/pydantic/pydantic/issues/11595.
